internal/authmiddleware/serverroute_auth.go

```diff
@@ -43,10 +39,76 @@ func (s *Server) handleAuth(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// Determine k8s username
-	k8sUID := user
-	k8sUsername := GetOidcUsername(s.config, preferredUsername)
-	k8sGroups := GetOidcGroups(s.config, splitGroups(groups))
+	// Authorization is required
+	if authHeader == "" {
+		s.logger.Error("Missing Authorization header")
+		http.Error(w, "Missing Authorization header", http.StatusUnauthorized)
+		return
+	}
+
+	// OIDCVerifier should always be initialized when /auth is enabled
+	if s.oidcVerifier == nil {
+		s.logger.Error("OIDC verifier is not initialized")
+		http.Error(w, "Internal server error: OIDC verifier not initialized", http.StatusInternalServerError)
+		return
+	}
+
+	// Extract token from Authorization header
+	token, err := ExtractBearerToken(authHeader)
+	if err != nil {
+		s.logger.Error("Failed to extract bearer token", "error", err)
+		http.Error(w, "Invalid Authorization header", http.StatusBadRequest)
+		return
+	}
+
+	// Verify the token with the OIDC provider
+	oidcClaims, isVerifyTokenFault, err := s.oidcVerifier.VerifyToken(r.Context(), token, s.logger)
+	if err != nil {
+		if isVerifyTokenFault {
+			// Server-side error (e.g., OIDC provider unavailable)
+			s.logger.Error("OIDC provider connection error", "error", err)
+			http.Error(w, "Internal server error: OIDC provider not available", http.StatusInternalServerError)
+			return
+		}
+
+		// Otherwise the token is invalid, reject with 400
+		s.logger.Error("OIDC token validation error", "error", err)
+		http.Error(w, "Invalid or expired OIDC token", http.StatusForbidden)
+		return
+	}
+
+	// Set the user identity variables from the OIDC token
+	k8sUID := oidcClaims.Subject
+	k8sUsername := GetOIDCUsernameFromToken(s.config, oidcClaims)
+	k8sGroups := GetOIDCGroupsFromToken(s.config, oidcClaims)
+
+	// Verify preferred username in header if available
+	if headerPreferredUsername != "" && k8sUsername != headerPreferredUsername {
+		s.logger.Error("Preferred username mismatch between token and headers",
+			"token preferred username", k8sUsername,
+			"header preferred username", headerPreferredUsername)
+		http.Error(w, "Username mismatch between token and headers", http.StatusUnauthorized)
+		return
+	}
+
+	// Verify UID in header if available
+	if headerUID != "" && k8sUID != headerUID {
+		s.logger.Error("UID mismatch between token and headers",
+			"token UID", k8sUID,
+			"header UID", headerUID)
+		http.Error(w, "UID verification failed", http.StatusUnauthorized)
+		return
+	}
+
+	// Verify groups in header if available
+	if len(headerGroups) > 0 {
```

Comment:
@JGuinegagne you menioned planning to replace header but still can be apploed to critical parts of auth flow: 

should this / critical parts of auth flow have tests?

Reply:
WDYM sorry? The `authmiddleware` no longer trusts these headers, in fact it throws a 401 if the header value doesn't match what's in the token. I kept them here just for auditability.